packages/event-listener/src/index.ts

```diff
@@ -1,128 +1,135 @@
-  handler: EventListenerOrEventListenerObject | null,
-  options?: AddEventListenerOptions
-];
+export type EventHandler<
+  EventMap extends
+    | HTMLElementEventMap
+    | WindowEventMap
+    | DocumentEventMap
+    | MediaQueryListEventMap
+    | Record<string, Event>,
+  EventName extends keyof EventMap
+> = (event: EventMap[EventName]) => void;
+
+export type EventListenerReturn = [stop: Fn, start: Fn];
 
-export type CreateEventListenerReturn = [
-  add: (el: EventTarget) => void,
-  remove: (el: EventTarget) => void
-];
+export type EventListenerDirectiveProps =
+  | [string, (e: any) => void, boolean | AddEventListenerOptions]
+  | [string, (e: any) => void];
 
 declare module "solid-js" {
   namespace JSX {
     interface Directives {
-      createEventListener: (
-        ref: HTMLElement,
-        props: Accessor<EventListenerProps<HTMLElement, {}>>
-      ) => [add: (target: EventTarget) => void, remove: (target: EventTarget) => void];
+      // directive types are very premissive to prevent type errors from incompatible types, since props cannot be generic
+      createEventListener: EventListenerDirectiveProps;
     }
   }
 }
 
 // only here so the `JSX` import won't be shaken off the tree:
 export type E = JSX.Element;
 
-type CreateEventListenerFn = <
-  EventMap extends Record<string, Event>,
-  Target extends EventTarget | EventTarget[]
->(
-  target: Target,
-  ...props: [Accessor<EventListenerProps<Target, EventMap>>] | EventListenerProps<Target, EventMap>
-) => CreateEventListenerReturn;
-
 /**
- * Creates an event listener helper primitive.
+ * Creates an event listener, that will be automatically disposed on cleanup.
  *
  * @param target - ref to HTMLElement, EventTarget or Array thereof
- * @param nameOrProps - name of the event or Accessor with the event props ([name, handler, options?])
- * @param handler - if nameOrProps contains a name, you can specify the handler here
- * @param options - if nameOrProps contains a name, you can specify event listener options
+ * @param eventName - name of the handled event
+ * @param handler - event handler
+ * @param options - addEventListener options
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener
+ * @see https://github.com/davedbase/solid-primitives/tree/main/packages/event-listener#createEventListener
  *
  * @example
- * ```tsx
- * const [add, remove] = createEventListener(
- *   document.getElementById("mybutton"),
- *   "mouseDown",
- *   () => console.log("Click")
- * );
- * // or as a directive
- * <MyButton use:createEventListener={() => ['click', () => console.log("Click")]}>Click!</MyButton>
- * // you can provide your own event map type:
- * createEventListener<{ myCustomEvent: Event }, Window>(window, 'myCustomEvent', () => console.log("yup!"));
- * ```
+ * const [stop, start] = createEventListener(
+ *   document.querySelector('.myEl'),
+ *   'click',
+ *   e => { ... }
+ * )
```

Comment:
Please provide the other examples as well. The previous example for the directive had a superfluous function wrapper, as the JSX expression is already wrapping its content.

Reply:
My idea here was that maybe in jsdoc we should place only a simple example, but provide a clickable link to the primitive readme with a deeper explanation. But that's true if this is supposed to show the API, even if briefly, then I should add the directive usage back in there.